OCaml/sudoku_solver/lib/sudoku.ml

```diff
@@ -0,0 +1,75 @@
+(* OCaml Version of the code *)
+
+module Sudoku = struct
+  (* Main function to run *)
+  type board = int array array
+
+  let print_board board = 
+    Array.iter (fun row ->
+      Array.iter (fun col -> Printf.printf "%d " col) row;
+      print_newline()) board
+  ;;
+
+  (* to check if the number is in row, col and subgrid *) 
+  let is_valid_move board row col num = 
+    Array.exists (( = ) num) board.(row) && 
+    Array.exists (fun r -> r.(col) = num) board &&
+    Array.exists (fun r -> Array.exists (( = ) num) (Array.sub r col 3)) (Array.sub board row 3)
+  ;;
+
+  (* check if each row and col if = 0, then return coordinates *)
+  let find_empty_cell board = 
+    let rec aux row col =
+      if row = Array.length board then None (* reach the end *)
+      else if col = Array.length board && row < Array.length board then aux (row + 1) col
+      else if board.(row).(col) = 0 then Some (row, col)
+      else aux row (col + 1)
+  in 
+    aux 0 0
+  ;;
+
+  let solve_sudoku board =
+    None  
+  ;;
+
+
+end
+
+  let%test "is_valid_move test" =
+    let board = [|
+      [|5; 3; 0; 0; 7; 0; 0; 0; 0|];
+    [|6; 0; 0; 1; 9; 5; 0; 0; 0|];
+    [|0; 9; 8; 0; 0; 0; 0; 6; 0|];
+    [|8; 0; 0; 0; 6; 0; 0; 0; 3|];
+    [|4; 0; 0; 8; 0; 3; 0; 0; 1|];
+    [|7; 0; 0; 0; 2; 0; 0; 0; 6|];
+    [|0; 6; 0; 0; 0; 0; 2; 8; 0|];
+    [|0; 0; 0; 4; 1; 9; 0; 0; 5|];
+    [|0; 0; 0; 0; 8; 0; 0; 7; 9|];  
+    |] in
+    let result =
+      Sudoku.is_valid_move board 0 2 4
+    && not (Sudoku.is_valid_move board 1 2 3)
+    && Sudoku.is_valid_move board 2 2 7
+    && Sudoku.is_valid_move board 3 4 5
+    && not (Sudoku.is_valid_move board 4 4 3)
+    in
+  Printf.printf "Test Result: %b\n" result;  (* Print the result *)
+  (*assert result; *)  (* Assert the result *)
+  ;;
```

Comment:
the `let%test` ppx swallows the output so the printing here does nothing I believe. I find ppx_tests not super good because they only tell you the function that failed but give you no way to debug. What you want instead are [`let%expect_test`](https://github.com/janestreet/ppx_expect) ppx tests instead which prints the diff for you.
```suggestion
  let%expect_test "is_valid_move test" =
    let board = [|
      [|5; 3; 0; 0; 7; 0; 0; 0; 0|];
    [|6; 0; 0; 1; 9; 5; 0; 0; 0|];
    [|0; 9; 8; 0; 0; 0; 0; 6; 0|];
    [|8; 0; 0; 0; 6; 0; 0; 0; 3|];
    [|4; 0; 0; 8; 0; 3; 0; 0; 1|];
    [|7; 0; 0; 0; 2; 0; 0; 0; 6|];
    [|0; 6; 0; 0; 0; 0; 2; 8; 0|];
    [|0; 0; 0; 4; 1; 9; 0; 0; 5|];
    [|0; 0; 0; 0; 8; 0; 0; 7; 9|];  
    |] in
    let result =
      Sudoku.is_valid_move board 0 2 4
    && not (Sudoku.is_valid_move board 1 2 3)
    && Sudoku.is_valid_move board 2 2 7
    && Sudoku.is_valid_move board 3 4 5
    && not (Sudoku.is_valid_move board 4 4 3)
    in
    Printf.printf "Test Result: %b\n" result;  (* Print the result *)
    [%expect {| |}]
  ;;
```

When you run `dune test` if there is a difference between what you expected to be printed, it will show in the terminal. In this case cos we define the expected result to be empty, it will show as an error and then you can inspect the result. If you think it looks correct, you can run `dune promote` and it will automatically update the expected result.

Btw `{||}` is how you'd write a string and use like the \<enter\> key to represent a new line. So 
```
{|blah
blah|}
```
means "blah\nblah" but formatted easier to see

Reply:
I see, in that case then, everything should be a separate test case?

When would be the case to use inline testing?

Reply:
Yes, I imagine that ppx_tests work best to just confirm some basic invariant. It makes sense to split them up into small cases so that even if you can't see what the output is failing, at least you don't need to dig into which assert is failing